File: backend/app/schemas/spiel.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict

from ..models.enums import ZusageStatus


class SpielCreate(BaseModel):
    mannschaft_id: int
    gegner: str
    ist_heimspiel: bool = True
    termin: datetime
    ort: Optional[str] = None
    externe_id: Optional[str] = None


class SpielUpdate(BaseModel):
    gegner: Optional[str] = None
    ist_heimspiel: Optional[bool] = None
    termin: Optional[datetime] = None
    ort: Optional[str] = None
    notiz: Optional[str] = None


class SpielRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    mannschaft_id: int
    gegner: str
    ist_heimspiel: bool
    termin: datetime
    ort: Optional[str]
    treffpunkt_ort: Optional[str]
    treffpunkt_zeit: Optional[datetime]
    notiz: Optional[str]
    aufstellung_freigegeben: bool


class TreffpunktUpdate(BaseModel):
    treffpunkt_ort: Optional[str] = None
    treffpunkt_zeit: Optional[datetime] = None
    notiz: Optional[str] = None


class AufstellungEintrag(BaseModel):
    position: int
    spieler_id: int
    ist_ersatz: bool = False


class AufstellungUpdate(BaseModel):
    eintraege: list[AufstellungEintrag]
    freigeben: bool = False


class AufstellungValidationResult(BaseModel):
    ok: bool
    warnungen: list[str] = []
    fehler: list[str] = []


class ZusageUpdate(BaseModel):
    status: ZusageStatus
    kommentar: Optional[str] = None


class ZusageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    spiel_id: int
    spieler_id: int
    status: ZusageStatus
    kommentar: Optional[str]
    geaendert_am: datetime


class PinZusageRequest(BaseModel):
    pin: str
    spieler_id: int
    status: ZusageStatus


class AufstellungEintragRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    position: int
    spieler_id: int
    ist_ersatz: bool


class SpielDetailRead(SpielRead):
    aufstellung: list[AufstellungEintragRead] = []
    zusagen: list[ZusageRead] = []


class ErsatzanfrageRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    spiel_id: int
    spieler_id: int
    position: int
    status: ZusageStatus
    erstellt_am: datetime

File: frontend/lib/models/spiel.dart
// Copyright (c) 2026 Noah, Luca, Sheila, Lando. All rights reserved.

enum ZusageStatus { offen, zugesagt, abgesagt }

ZusageStatus zusageStatusFromString(String? s) {
  switch (s) {
    case 'zugesagt':
      return ZusageStatus.zugesagt;
    case 'abgesagt':
      return ZusageStatus.abgesagt;
    default:
      return ZusageStatus.offen;
  }
}

String zusageStatusToApi(ZusageStatus s) => s.name;

class Zusage {
  final int id;
  final int spielId;
  final int spielerId;
  final ZusageStatus status;
  final String? kommentar;

  Zusage({
    required this.id,
    required this.spielId,
    required this.spielerId,
    required this.status,
    this.kommentar,
  });

  factory Zusage.fromJson(Map<String, dynamic> j) => Zusage(
        id: j['id'] as int,
        spielId: j['spiel_id'] as int,
        spielerId: j['spieler_id'] as int,
        status: zusageStatusFromString(j['status'] as String?),
        kommentar: j['kommentar'] as String?,
      );
}

class AufstellungEintrag {
  final int position;
  final int spielerId;
  final bool istErsatz;

  AufstellungEintrag({
    required this.position,
    required this.spielerId,
    this.istErsatz = false,
  });

  factory AufstellungEintrag.fromJson(Map<String, dynamic> j) => AufstellungEintrag(
        position: j['position'] as int,
        spielerId: j['spieler_id'] as int,
        istErsatz: j['ist_ersatz'] as bool? ?? false,
      );

  Map<String, dynamic> toJson() => {
        'position': position,
        'spieler_id': spielerId,
        'ist_ersatz': istErsatz,
      };
}

class Spiel {
  final int id;
  final int mannschaftId;
  final String gegner;
  final bool istHeimspiel;
  final DateTime termin;
  final String? ort;
  final String? treffpunktOrt;
  final DateTime? treffpunktZeit;
  final String? notiz;
  final bool aufstellungFreigegeben;
  final List<AufstellungEintrag> aufstellung;
  final List<Zusage> zusagen;

  Spiel({
    required this.id,
    required this.mannschaftId,
    required this.gegner,
    required this.istHeimspiel,
    required this.termin,
    this.ort,
    this.treffpunktOrt,
    this.treffpunktZeit,
    this.notiz,
    this.aufstellungFreigegeben = false,
    this.aufstellung = const [],
    this.zusagen = const [],
  });

  factory Spiel.fromJson(Map<String, dynamic> j) => Spiel(
        id: j['id'] as int,
        mannschaftId: j['mannschaft_id'] as int,
        gegner: j['gegner'] as String,
        istHeimspiel: j['ist_heimspiel'] as bool,
        termin: DateTime.parse(j['termin'] as String),
        ort: j['ort'] as String?,
        treffpunktOrt: j['treffpunkt_ort'] as String?,
        treffpunktZeit: j['treffpunkt_zeit'] != null
            ? DateTime.parse(j['treffpunkt_zeit'] as String)
            : null,
        notiz: j['notiz'] as String?,
        aufstellungFreigegeben: j['aufstellung_freigegeben'] as bool? ?? false,
        aufstellung: (j['aufstellung'] as List? ?? [])
            .map((e) => AufstellungEintrag.fromJson(e as Map<String, dynamic>))
            .toList(),
        zusagen: (j['zusagen'] as List? ?? [])
            .map((e) => Zusage.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}
